extensions.c

```diff
@@ -40,6 +42,94 @@ struct hibaext {
 	struct pair pairs;
 };
 
+static struct sshbuf*
+hibaext_maybe_inflate(struct sshbuf *extensions) {
+	int ret = HIBA_OK;
+	struct sshbuf *decomp = NULL;
+#ifdef WITH_EXTENSION_COMPRESSION
+	size_t len = sshbuf_len(extensions);
+	const unsigned char *data = sshbuf_ptr(extensions);
+	z_stream stream;
+
+	memset(&stream, 0, sizeof(stream));
+
+	inflateInit(&stream);
+	stream.next_in = data;
+	stream.avail_in = len;
+
+	do {
+		unsigned char buf[512];
+
+		stream.next_out = buf;
+		stream.avail_out = sizeof(buf);
+		ret = inflate(&stream, Z_SYNC_FLUSH);
+		if (ret != Z_OK && ret != Z_STREAM_END) {
+			break;
+		}
+		if (decomp == NULL) {
+			decomp = sshbuf_new();
+		}
+		sshbuf_put(decomp, buf, stream.total_out);
+	} while (ret != Z_STREAM_END);
+	inflateEnd(&stream);
+
+	if (ret == Z_STREAM_END) {
+		debug2("hibaext_maybe_inflate: using decompressed extension %zu -> %zu bytes", len, sshbuf_len(decomp));
+		return decomp;
+	}
+	sshbuf_free(decomp);
+#endif /* WITH_EXTENSION_COMPRESSION */
+	return sshbuf_fromb(extensions);
+}
+
+static int
+hibaext_maybe_deflate(struct sshbuf *in, int compress, struct sshbuf *out) {
+	int ret = HIBA_OK;
+	struct sshbuf *temp = in;
+
+#ifdef WITH_EXTENSION_COMPRESSION
+	if (compress) {
+		z_stream stream;
+
+		temp = sshbuf_new();
+		memset(&stream, 0, sizeof(stream));
+
+		deflateInit(&stream, Z_BEST_COMPRESSION);
+		stream.next_in = (unsigned char*)sshbuf_ptr(in);
+		stream.avail_in = sshbuf_len(in);
+
+		debug3("hibaext_maybe_deflate: compression in %d.", stream.avail_in);
+		while (ret != Z_STREAM_END) {
+			unsigned char buf[512];
+
+			stream.next_out = buf;
+			stream.avail_out = sizeof(buf);
+			ret = deflate(&stream, Z_FINISH);
+			if (ret != Z_OK && ret != Z_STREAM_END) {
+				debug2("hibaext_maybe_deflate: deflate returned %d", ret);
+				ret = HIBA_INTERNAL_ERROR;
+				goto err;
+			}
+			sshbuf_put(temp, buf, stream.total_out);
+		}
+		ret = HIBA_OK;
+		debug3("hibaext_maybe_deflate: compression out %zu.", sshbuf_len(temp));
+err:
+		deflateEnd(&stream);
+	}
+#endif  /* WITH_EXTENSION_COMPRESSION */
+
+	if (ret == HIBA_OK) {
+		if (sshbuf_putb(out, temp) < 0) {
+			ret = HIBA_INTERNAL_ERROR;
+		}
+	}
+	if (temp != in) {
+		sshbuf_free(temp);
+	}
+	return ret;
+}
+
 int
 hibaext_decode(struct hibaext *ext, struct sshbuf *blob) {
```

Comment:
Can we make blob a const here? Just to make it more explicit that it's a read-only input parameter.